### PR TITLE
Beginning of 0.10.x support - timestamps

### PR DESCRIPTION
**Pull request by mrb**
This Pull Request begins the work to support Kafka version 0.10.x, which has protocol changes outlined here: https://cwiki.apache.org/confluence/display/KAFKA/KIP-32+-+Add+timestamps+to+Kafka+message

I'm opening this PR now as per conversations with @dasch

### CODE REVIEW

**Comment by pbrisbin** [comment]
Nice!

I doesn't look like there are any changes to `FetchRequest`/`FetchResponse` here. Does that mean we'll fall into Phase 2, scenario 2.c [here](https://cwiki.apache.org/confluence/display/KAFKA/KIP-32+-+Add+timestamps+to+Kafka+message#KIP-32-AddtimestampstoKafkamessage-Compatibility,Deprecation,andMigrationPlan):

> When broker sees a fetch request V1 (Supporting MessageAndOffset = V0), because the data format on disk is MessageAndOffset V1, it will NOT use the zero-copy transfer. Instead the broker will read the message from disk, down-convert them to V0 and reply using fetch response V1 with MessageAndOffset V0.

File: lib/kafka/fetched_message.rb [hunk 1]
```diff
@@ -16,12 +16,16 @@ class FetchedMessage
     # @return [Integer] the offset of the message in the partition.
     attr_reader :offset
 
-    def initialize(value:, key:, topic:, partition:, offset:)
+    # @return [Integer] the timestamp associated with the message.
+    attr_reader :timestamp
+
+    def initialize(value:, key:, topic:, partition:, offset:, timestamp:nil)
```
Comment (pbrisbin) [inline lib/kafka/fetched_message.rb]: 💄 `timestamp: nil` (missing space)

**Comment by dasch** [comment]
@mrb should this be closed in favor of #279?

**Comment by mrb** [comment]
@dasch I'll take a look! I don't believe it handles both decoding cases, and I have some more code to push, and then we can decide.

**Comment by mrb** [comment]
@dasch I'm pretty happy with this PR in its current state - going to start testing with our app today! Happy to just use the other PR or combine them somehow if that's easiest. /c @pbrisbin

**Comment by dasch** [comment]
It would be nice to have a single PR. Also, I think this PR needs to handle timestamps being in milliseconds in Kafka.

**Comment by mrb** [comment]
@dasch Gotcha. I've since addressed the milliseconds issue.

File: lib/kafka/protocol/message.rb [hunk 2]
```diff
@@ -81,7 +104,18 @@ def self.decode(decoder)
         new(key: key, value: value, codec_id: codec_id, offset: offset)
       end
 
-      private
+      def self.v2_decode(offset, message_decoder)
+        attributes = message_decoder.int8
+        timestamp = message_decoder.int64
```
Comment (dasch) [inline lib/kafka/protocol/message.rb]: The timestamp should probably be 1) converted from ms to s, and 2) wrapped in a Time object.

File: spec/spec_helper.rb [hunk 3]
```diff
@@ -1,3 +1,10 @@
+require 'simplecov'
+SimpleCov.start do
+  add_filter do |src|
+    src.filename =~ /spec/
+  end
+end
```
Comment (dasch) [inline spec/spec_helper.rb]: This seems unrelated...

**Comment by dasch** [comment]
@yingnanliu can you and @mrb work out a single, shared PR for this feature? I'm going to end up closing one of them.

:heart:

**Comment by yingnanliu** [comment]
@dasch Of course, we'll discuss how to merge these two PRs.
Hi @mrb, I see a major difference between these two PRs is that #278 supports decoding for both V1 and V2. IIUC, the `FetchRequest` is already set to V2, and then brokers will send V2 responses. V1 decoding is not needed in this case. What do you think?

**Comment by dasch** [comment]
@yingnanliu the reason for that is that the v2 fetch API can still serve up messages written in the v1 format – the "magic byte" of the encoded message dictates the format of the message itself, regardless of the fetch API version used.

**Comment by mrb** [comment]
@dasch @yingnanliu It seemed worth preserving that backward compatibility here. It's very similar to how the official Java client handles it.

**Comment by pbrisbin** [comment]
When performing the update from 0.8/0.9 to 0.10 it's important to keep the log message format at the older version if there are still any older clients around. If you move directly to the 0.10 format, there have been reports of large performance issues on the brokers since they have to down-convert the messages for any older clients and can't use zero-copy transfer.

http://kafka.apache.org/documentation.html#upgrade_10_performance_impact

I _believe_ this is the V1/V2 you're talking about, which means that until we've upgrade all clients and moved the log message format up to 0.10, the clients that _are_ upgraded will have to handle V1 messages.

**Comment by dasch** [comment]
@pbrisbin I believe even after the protocolversion has been set to 0.10 you'll still have to deal with messages in the old format – otherwise Kafka would have to rewrite the entire history, as the message sets are copied straight from disk to the wire under normal operation.

**Comment by pbrisbin** [comment]
@dasch thanks for that clarification, that makes sense. Our data is pretty ephemeral so that's unlikely to be the case for us, but would be for anyone reading their data long-term.

Note: I think we should careful about saying "protocol version" vs "log message version" since those are two independent settings, one for inter-broker communication and one for log storage (which impacts broker-client communication).

File: lib/kafka/protocol/message.rb [hunk 4]
```diff
@@ -81,7 +104,18 @@ def self.decode(decoder)
         new(key: key, value: value, codec_id: codec_id, offset: offset)
       end
 
-      private
+      def self.v2_decode(offset, message_decoder)
+        attributes = message_decoder.int8
+        timestamp = Time.at(message_decoder.int64/1000)
```
Comment (asoules) [inline lib/kafka/protocol/message.rb]: Should this divide by the float `1000.0` to preserve milliseconds?

**Comment by dasch** [comment]
Closing in favor of #376.
